Addition of support for ADLX312 and ADXL314 for the ADXL313 driver.

danmois opened the pull request:
ADXL312 and ADXL314 are small, thin, low power, 3-axis accelerometers with high resolution (13-bit) measurement up to +/-12 g and +/- 200 g, respectively. They are based on the same ASIC as ADXL313 and support for their operation is added to the already existing ADXL313 driver.

buha commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -156,12 +197,33 @@ static int adxl313_read_raw(struct iio_dev *indio_dev,
 		*val = 0;
-		*val2 = 9576806;
+
+		if (data->id->driver_data == ADXL312) {
```
instead of doing these checks here at each _read_raw() it's more efficient to do them in _probe() and store the scale factor in driver structure, and here you'd just:
`*val = data->scale_factor;`

danmois replied:
Thank you for the observation, I added the scale_factor member.

buha commented on drivers/iio/accel/adxl313_spi.c:
```diff
@@ -51,26 +71,40 @@ static int adxl313_spi_probe(struct spi_device *spi)
 		return ret;
 
-	regmap = devm_regmap_init_spi(spi, &adxl313_spi_regmap_config);
+	if (id->driver_data == 0) {
```
instead of all these ifs you could have a table indexed by the device type where you store the regmap_config:
`regmap = devm_regmap_init_spi(spi, table[id->driver_data]);'

danmois replied:
devm_regmap_init_spi has an element of an array of regmaps as the second parameter.

buha commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -242,41 +308,55 @@ static int adxl313_setup(struct device *dev, struct adxl313_data *data,
-	if (ret)
-		return ret;
+	/* Sets the range to maximum, full resolution, for ADXL312 and ADXL313 */
+	if (dev_type != ADXL314) {
```
future proof it with `dev_type == ADXL312 || dev_type == ADXL313`

danmois replied:
The condition has been changed to the suggested format.

danmois commented:
V1
-  added the scale_factor member for driver data
- added an array for regmaps associated to every device
- replaced the condition for future proofing

danmois commented:
V2
- fixed cppcheck errors and warnings

danmois commented:
V2
- implemented cppcheck suggested fixes

amiclaus commented on drivers/iio/accel/adxl313.h:
```diff
@@ -37,18 +38,28 @@
+		       const struct spi_device_id *id,
 		       int (*setup)(struct device *, struct regmap *));
-#endif /* _ADXL313_H_ */
+#endif /* _ADXL313_H_ */
```
nit: unrelated change.

amiclaus replied:
nit: unrelated changes

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -242,41 +286,55 @@ static int adxl313_setup(struct device *dev, struct adxl313_data *data,
-	if (regval != ADXL313_DEVID0) {
-		dev_err(dev, "Invalid manufacturer ID: 0x%02x\n", regval);
+	if (dev_type == ADXL313 && regval != ADXL313_DEVID0) {
+		dev_err(dev, "Invalid manufacturer ID: %#02x\n", regval);
```
was this change done on purpose? looks like unrelated change to the commit subject.

danmois replied:
True, unrelated change, I removed the change.

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -242,41 +286,55 @@ static int adxl313_setup(struct device *dev, struct adxl313_data *data,
+		ret = regmap_update_bits(data->regmap, ADXL313_REG_DATA_FORMAT,
+					ADXL313_RANGE_MSK,
+					FIELD_PREP(ADXL313_RANGE_MSK, 0x02));
+
```
nit: extra line

danmois replied:
Thank you for the review, all signaled problems have been solved.

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -198,9 +238,10 @@ static int adxl313_write_raw(struct iio_dev *indio_dev,
+		 * 8-bit resolution at minimum range, that is 4x accel data scale
 		 * factor at full resolution
 		 */
+
```
nit: unrelated change

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -309,23 +367,50 @@ int adxl313_core_probe(struct device *dev,
 	data->regmap = regmap;
+	data->id = id;
+
+	if (id->driver_data == ADXL312) {
```
nit: redundant brackets

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -309,23 +367,50 @@ int adxl313_core_probe(struct device *dev,
+		data->scale_factor = 28425072;
+	}
+
+	if (id->driver_data == ADXL313) {
```
ditto

danmois commented:
V2

- implemented cppcheck suggested fixes

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -309,23 +367,50 @@ int adxl313_core_probe(struct device *dev,
+		data->scale_factor = 9576806;
+	}
+
+	if (id->driver_data == ADXL314) {
```
ditto

amiclaus commented on drivers/iio/accel/adxl313_core.c:
```diff
@@ -309,23 +367,50 @@ int adxl313_core_probe(struct device *dev,
+	}
+
+	if (id->driver_data == ADXL313) {
+		/*
```
keep comments consistent: add device nice as you did for adxl314

danmois replied:
Comments added

danmois commented:
V3
- implemented review comments (removed unrelated change, extra lines, unnecessary brackets)

amiclaus reviewed (commented):
I think these patches can be sent upstream for further review.

nunojsa commented:
> I think these patches can be sent upstream for further review.

Surely the way to go in here. Please send these upstream before merging them in our tree.

danmois commented:
V4

- variant accepted upstream
